src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelOperationManager.cs

```diff
@@ -101,66 +114,111 @@ private bool RunWorkInParallel()
         if (_runWorkload == null)
             throw new InvalidOperationException($"{nameof(_runWorkload)} was not provided.");
 
-        // Reserve slots and assign them work under the lock so we keep
-        // the slots consistent.
-        List<SlotWorkloadPair> workToRun = new();
+        // Reserve slots and assign them work under the lock so we keep the slots consistent.
+        Slot[] slots;
         lock (_lock)
         {
-            if (_workloads.Count == 0)
-                return false;
-
             // When HandlePartialDiscovery or HandlePartialRun are in progress, and we call StopAllManagers,
             // it is possible that we will clear all slots, and have RunWorkInParallel waiting on the lock,
             // so when it is allowed to enter it will try to add more work, but we already cancelled,
             // so we should not start more work.
             if (!_acceptMoreWork)
                 return false;
 
-            var availableSlots = _managerSlots.Where(slot => slot.IsAvailable).ToList();
-            var availableWorkloads = _workloads.Where(workload => workload != null).ToList();
-            var amount = Math.Min(availableSlots.Count, availableWorkloads.Count);
-            var workloadsToRun = availableWorkloads.Take(amount).ToList();
-
+            // We grab all empty slots.
+            var availableSlots = _managerSlots.Where(slot => !slot.HasWork).ToImmutableArray();
+            var occupiedSlots = MaxParallelLevel - (availableSlots.Length - PreStartCount);
+            // We grab all available workloads.
+            var availableWorkloads = _workloads.Where(workload => workload != null).ToImmutableArray();
+            // We take the amount of workloads to fill all the slots, or just as many workloads
+            // as there are if there are less workloads than slots.
+            var amount = Math.Min(availableSlots.Length, availableWorkloads.Length);
+            var workloadsToAdd = availableWorkloads.Take(amount).ToImmutableArray();
+
+            // We associate each workload to a slot, if we reached the max parallel
+            // level, then we will run only initalize step of the given workload.
             for (int i = 0; i < amount; i++)
             {
                 var slot = availableSlots[i];
-                slot.IsAvailable = false;
-                var workload = workloadsToRun[i];
-                workToRun.Add(new SlotWorkloadPair(slot, workload));
+                slot.HasWork = true;
+                var workload = workloadsToAdd[i];
+                slot.ShouldPreStart = occupiedSlots + i + 1 > MaxParallelLevel;
```

Comment:
Check all properties are thread safe.

Reply:
Is that necessary? All changes to non-active slots are done in the same lock, either here, or in handling partial complete. Or is there a big hole in my thinking that you found?

Reply:
Closing this as discussed offline.